e2e/nomostest/portforwarder/port_forwarder.go

```diff
@@ -110,32 +118,72 @@ func NewPortForwarder(
 func (pf *PortForwarder) LocalPort() (int, error) {
 	pf.mux.Lock()
 	defer pf.mux.Unlock()
-	if pf.localPort == 0 {
+	if pf.subprocess == nil {
 		err := pf.start()
-		return pf.localPort, err
+		return pf.subprocess.localPort, err
+	} else if !pf.ready {
+		// TODO: block until ready?
+		return 0, errors.Errorf("port-forward not ready to deployment %s/%s",
+			pf.ns, pf.deployment)
 	}
-	return pf.localPort, nil
+	return pf.subprocess.localPort, nil
 }
 
-func (pf *PortForwarder) setPort(cmd *exec.Cmd, port int, pod string, async bool) {
-	pf.logger.Infof("updating port-forward %s:%d -> %s:%d", pf.pod, pf.localPort, pod, port)
-	if async {
-		pf.mux.Lock()
-	}
-	pf.cmd = cmd
-	pf.localPort = port
-	pf.pod = pod
-	if async {
-		pf.mux.Unlock()
+// updateSubprocessAsync calls updateSubprocess with the lock
+func (pf *PortForwarder) updateSubprocessAsync(proc *subprocess, ready bool) {
+	pf.mux.Lock()
+	defer pf.mux.Unlock()
+	pf.updateSubprocess(proc, ready)
+}
+
+// updateSubprocess handles changes to the subprocess
+func (pf *PortForwarder) updateSubprocess(proc *subprocess, ready bool) {
+	switch {
+	case ready:
+		// Subprocess become ready
+		pf.logger.Infof("updating port-forward localhost:%d -> %s:%d (ready: %v)", proc.localPort, proc.pod, pf.port, ready)
+		pf.subprocess = proc
+		pf.ready = true
+		if pf.onReadyFunc != nil {
+			pf.onReadyFunc(proc.localPort, proc.pod)
```

Comment:
This will cause a deadlock if we enter `updateSubprocess` with the lock, and the `onReadyFunc` calls `LocalPort` and tries to re-obtain the lock (which it currently does for the git PortForwarder).

I confirmed after testing this locally that this leads to a deadlock.

Reply:
Good catch. How about we launch the callback in a goroutine?

We need waitAndUpdateSubprocess to run in the background anyway, while the callback is running, in case the port-forward stops while the callback is running. 

It occurs to me, that if the callback fails, we want to keep going anyway, because the next callback might work. But for that to work, we also need to give the callback its own lock, to prevent multiple running simultaneously.

Reply:
My first iteration was to avoid calling `LocalPort` from inside the callback, and the second iteration was to invoke the callback outside of the lock. I think either of those approaches work without too much added complexity

Reply:
I think running the callback async would work but adds some unnecessary complexity

Reply:
It is more complex to run in a goroutine, but it should allow us to refactor the callback to allow it to fail and retry with a new port-forward. Unfortunately, that requires refactoring Repository first to return errors, instead of failing the test.